Cache Pure Top Level Definitions on startup

ChrisPenner opened the pull request:
## Overview

Before this change all pure top-level definitions were evaluated at every call site, which is obviously unnecessary.

Now we can detect these using type info from the codebase, and inline the evaluated result into usage sites.

We also discussed whether to bake the evaluated results into the compiled artifact.
The consensus was yes, to serialize the results; I'd still like to do that, but after a day of fiddling it proved tougher than I expected.
Issues with that were:

1. Closures can contain `Foreign`s, which are arbitrary Haskell values, most of these won't appear in a pure value, but some thing like stdin/stdout Handles do, patterns as well, and things like text/bytes literals. These are all serializable in their own way, but it takes some care to ensure we do it correctly.
2. Closures contain argument references which refer to closures via `Seg 'BX`, but this is fixed in the Stack implementation to include inlined Sections, which we can't serialize. We can most likely either parameterize this type, or replace usage sites with a more generic version, but it affects a big chunk of things and is going to be a bit of a pain to figure out and propagate.

So serializing the pre-evalutated pure results is possible with more work, but it'd be nice to release the non-serializing version for now. The currently implemented version pre-evaluates cacheable values on executable startup.

When running something from a cold runtime (e.g. via `ucm run` or a fresh `ucm` session) there'll be a short delay while ucm evaluates everything, but it'll be snappy after that.

## Implementation notes

* I also did the optimization Dan suggested to avoid the `RComb` indirection by including the CombIx directly in the Section/Instr and avoid an additional redirection. This will also help with serializing closures, but as mentioned above that part isn't done yet.
* Add `getTypeOfTerm` to Runtime CodeLookup.

It's a bit annoying, because in order to evaluate the pure defns I need the rest of all the combs to be in the right format for the runtime, so here's how it goes:

1. Inline new combs
2. Iterate over new cacheable definitions and evaluate them, replacing the comb in the map with the evaluated closure
3. Re-inline new combs so they now inline the evaluated results.

## Benchmarks

Made a test which references a 1k element map definition:

```
map1K = range 0 1000 |> map (i -> (i, i)) |> Map.fromList

pureTest : '{IO, Exception} ()
pureTest =
  do
    use List map replicate
    use Nat + - == range
    repeat n action =
      if n == 0 then ()
      else
        _ = action()
        repeat (n - 1) action

    printTime
      "Map.lookup (1k element map)" 1 let
        n -> repeat n do Map.get 12 map1K
```

```
=======Trunk
Map.lookup (1k element map)
3.361018ms

=======This Branch
Map.lookup (1k element map)
5.3µs
```

## Loose ends

See overview about serializing closures.

ChrisPenner committed:
Handle cached closures in the Machine

ChrisPenner committed:
Add closures as a GComb constructor

ChrisPenner committed:
Compiling, but still need to actually inline closures

ChrisPenner committed:
Attempt to pre-eval

ChrisPenner committed:
Edit CodeLookup

ChrisPenner committed:
Compiling somehow

ChrisPenner committed:
Check types of refs right before passing to CCache

ChrisPenner committed:
Don't thread cacheability through floating

ChrisPenner committed:
Use backmap to look up types of codebase refs of top level defs

ChrisPenner committed:
Add type aliases for refs

ChrisPenner committed:
Debugging info

ChrisPenner committed:
Working pre-evaluated closures.

ChrisPenner committed:
Cleanup

ChrisPenner committed:
Fix missing pattern matches on Clos's

ChrisPenner committed:
Store srcCombs in SCache

ChrisPenner committed:
Pre-eval constants when loading from .uc files

ChrisPenner committed:
Start on serializing closures

ChrisPenner committed:
Serialization WIP

ChrisPenner committed:
Split CombIx out of RComb

ChrisPenner committed:
Successfully split of CombIx

ChrisPenner committed:
Handle serializing/deserializing split up combs

ChrisPenner committed:
Resolve conflicts

ChrisPenner committed:
Rewrite pre-evaluation

ChrisPenner committed:
Fixed closure embedding

ChrisPenner committed:
Don't serialize Closures

ChrisPenner committed:
Serialize cacheable combs and re-eval on load

ChrisPenner committed:
automatically run ormolu

ChrisPenner committed:
PR Cleanup

ChrisPenner commented on unison-runtime/src/Unison/Runtime/Machine.hs:
```diff
@@ -4,14 +4,13 @@
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE ViewPatterns #-}
--- TODO: Fix up all the uni-patterns
-{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
```
I replaced the uni-patterns with explicit errors so I could re-enable this.

ChrisPenner commented on unison-runtime/src/Unison/Runtime/Machine.hs:
```diff
@@ -129,19 +156,23 @@ baseCCache sandboxed = do
-    combs :: EnumMap Word64 RCombs
-    combs =
-      ( mapWithKey
+    srcCombs :: EnumMap Word64 Combs
```
I now include all the original non-recursive Combs in the CCache now because it makes it much easier to serialize

ChrisPenner commented on unison-runtime/src/Unison/Runtime/Machine.hs:
```diff
@@ -1833,23 +1879,26 @@ yield !env !denv !activeThreads !ustk !bstk !k = leap denv k
+          ustk <- ensure ustk uf
+          bstk <- ensure bstk bf
+          eval env denv activeThreads ustk bstk k ref nx
+        CachedClosure _w _clo -> do
+          error "TODO: Get help from Dan"
```
Just need a hand from @dolio  on this bit, surprisingly all my tests so far have run without it 😆

dolio replied:
I refactored `MCode` `Let` and associated stuff so that this case no longer occurs. The combinator in a `Let` was meant to be the resumption to the middle of a function, so it makes no sense to have a cached value there. The combinator was only useful for doing a second stack protection check in case we're resuming via a captured continuation.

Might even speed things up a bit due to less indirection.

dolio committed:
Tweak MCode Let representation

This introduces a minor change to the structure of MCode Let
bindings. Before various refactoring, a Let would use a CombIx to
indicate where execution should resume once the binding is
evaluated. Now the code is resolved ahead of time and referred to
directly. However, logically this code is just a further Section;
the rest of the actual function that contained the Let.

The purpose of using a Comb was to store stack protection
information, so that the stack check can be performed upon reentry.
This is necessary if the continuation has been captured and we are
resuming in a context that was not the original entry point of the
resumption.

Now, though, a top level 'comb' can be either code or a memoized
value, but a resumption will never be the latter. So Let has been
changed to store the relevant information, instead of delegating to
Comb.

dolio committed:
Add some documentation about the new MCode Let

dolio committed:
Remove now unnecessary prettyIx

dolio committed:
automatically run ormolu

dolio committed:
Fix unused binding warning

dolio committed:
Merge remote-tracking branch 'refs/remotes/origin/cp/cache-toplevel' into cp/cache-toplevel

ChrisPenner committed:
Re-merge trunk and fix recursion schemes stuff

pchiusano reviewed (approved):
🎉

dolio commented on unison-runtime/src/Unison/Runtime/Machine.hs:
```diff
@@ -732,38 +775,43 @@ apply ::
-          (useg, bseg) <- closeArgs C ustk bstk useg bseg args
-          ustk <- discardFrame =<< frameArgs ustk
-          bstk <- discardFrame =<< frameArgs bstk
+apply !env !denv !activeThreads !ustk !bstk !k !ck !args = \case
```
I think it doesn't really make much sense to have a partial application to a `CachedClosure`, similar to the `Let` situation. I haven't finished reading everything, but maybe there needs to be a type that is actually just combinator references, not both those and pre-evaluated closures.

dolio reviewed (changes requested):
So, I noticed one thing that seems weird to me. Having a cached closure in a `PAp` seems wrong, so I think we need to tweak something. I'm not sure what yet, though. I'll think about it and look things over some more.

dolio commented:
Okay, so, maybe it's as simple as this...

Separate the `Lam` constructor of `GComb` into its own type. This is the type that should occur in `PAp`.

Make `GComb` a sum referring to this new type. The new type should be unpacked into `GComb`, and pattern aliases could make using it look essentially like the current `GComb`.

Occurrences of the new, separate type don't need to be lazy, because they aren't from knot tying, they're the concrete entry info for an actual combinator.

dolio committed:
Factor GComb a bit, and make PAp more correct

A PAp should only contain an actual combinator, not a cached value.
So, the combinator case has been factored out of and unpacked into
GComb. This way a PAp can refer to the factored-out part.

dolio committed:
automatically run ormolu

ceedubs commented:
@ChrisPenner I'm curious about when this kicks in and what "on startup" means. Could you help me understand? Some of my questions:

- You've specified that this applies to `ucm run`. Does it apply to `run.compiled` and `run.native` as well?
- For `run`, `run.compiled`, and `run.native`, does it only evaluate/cache the dependencies of the function being run?
- It sounds like this also applies to interactive ucm sessions. Does it cache the result of every pure top-level term in the codebase at startup? Or when you switch to a project does it compute only the ones with names in the current branch of the project? Or something else entirely?
- Is this relevant for both the interpreter and the JIT runtime?

Here's one case that I'm wondering about:

Historically it hasn't been too bad to have a (pure) test in a project that checks 100k samples and takes 10 seconds to evaluate, because the result gets cached. But are we going to pay that 10 second penalty just to cache `[Ok "passed"]` in more places now (like when I run my `main` method or when I switch into that project)?

ceedubs commented:
I should have mentioned in my last comment that I'm excited about this change @ChrisPenner! I've seen repeated evaluation bite newcomers, and I've been curious about to what extent it might be hitting us with certain `Pattern` values, decoders, etc. Thank you!

dolio committed:
Include cacheability information in Code values

This allows code sent between machines to pre-evaluate things if
applicable. Naturally, this requires serialization version changes.
Some tweaks have been made to avoid changing hashes as much as
possible between versions.

Old serialized values and code can still be loaded, but obviously
they will be treated as completely uncacheable.

dolio committed:
Merge remote-tracking branch 'refs/remotes/origin/cp/cache-toplevel' into cp/cache-toplevel

dolio committed:
automatically run ormolu

dolio committed:
Fix MCode tests

ChrisPenner commented:
@ceedubs 

> For run, run.compiled, and run.native, does it only evaluate/cache the dependencies of the function being run?

It will run and cache each pure top-level definition as part of adding it to the runtime's code-cache, we only load the dependencies of whatever we're trying to run, so it'll evaluate dependencies of a `main` the _first_ time it's added to the runtime's code cache (e.g. once per ucm session, or once at the start of a `run.compiled` when we load the code from a `uc` file.)

> You've specified that this applies to ucm run. Does it apply to run.compiled and run.native as well?

See above, for `ucm run` and `run.compiled` the pre-evaluation triggers for each thing the first time it's added to the code cache.

`run.native` doesn't benefit from this change at all unfortunately.


> It sounds like this also applies to interactive ucm sessions. Does it cache the result of every pure top-level term in the codebase at startup? Or when you switch to a project does it compute only the ones with names in the current branch of the project? Or something else entirely?

It only pre-evaluates the dependencies of things you try to `run`, and will cache evaluation results for the lifespan of that runtime, which corresponds to the lifetime of the UCM process for the most part, e.g. if you `run main` it'll evaluate all cacheable definitions which `main` depends on (transitively). If you `run main` again, all the code is already loaded, so it won't do it again. If you change `main` it'll load any new hashes and evaluate anything cacheable in ONLY the new hashes, old ones are still cached.

> Is this relevant for both the interpreter and the JIT runtime?
Only the interpreter for now

> Historically it hasn't been too bad to have a (pure) test in a project that checks 100k samples and takes 10 seconds to evaluate, because the result gets cached. But are we going to pay that 10 second penalty just to cache [Ok "passed"] in more places now (like when I run my main method or when I switch into that project)?

This is only an issue if you reference the test in your `main`.

We may want to look into optimizations for docs and termLinks where we don't necessarily depend on the actual result of a referenced definition 🤔

ChrisPenner commented:
One downside of this change is if you have code like this:

```
main = do
  if something 
    then exit 0
    else Map.lookup k hugeExpensiveMap
```

Previously `hugeExpensiveMap` would only be evaluated when it was actually hit by the interpreter, so this program would exit immediately if `something` is true. 
Now we'll pre-evaluate `hugeExpensiveMap` once before even starting to run the program, which could lead to an increased startup time in some cases.

So this change will be good for longer-term processes like cloud and webservers, but potentially worse for short-lived processes like cli apps.

If the problem proves noticeable enough we probably can fix this delay by continuing on the Closure serialization work I mention in the overview, which I think we still want to do, the work was just starting to hit diminishing returns for how tricky it was when we have bigger possible gains with other work 😄 .

pchiusano commented:
I noticed this broke the serialized form round trip golden tests. Which might be fine, since I'd expect this could change the serialized form for some definitions, but it's worth checking. 

@SystemFw do we have Unison Cloud client tests to make sure that no hashes change for schema types and functions? The way these could work is to just `crypto.hash` various values and save as a hex string literal, and then the test does a comparison and fails if the hashes differ.

SystemFw commented:
We now have a test in the Cloud client to test for the relevant hash changes. It passes on this branch.

dolio committed:
Fix a guard in unison-runtime

It was meant to be a test in a `match` expression, but was missing
a #:when

dolio committed:
Bump @unison/internal version to fix jit compatibility for serialization

dolio committed:
Merge remote-tracking branch 'origin/trunk' into cp/cache-toplevel

dolio committed:
Bump runtime-tests version for new serialization format

aryairani commented:
> serializing the pre-evaluated pure results is possible with more work

fwiw, contrary to the majority opinion in the meeting, I don't think this would be a good idea. afaik other languages don't do it, apart maybe from some trivial constant folding — if someone really wants to precompute a pure result, they can always use `add.run`

aryairani commented on .github/workflows/ci.yaml:
```diff
@@ -14,8 +14,8 @@ on:
   ormolu_version: 0.7.2.0
-  jit_version: "@unison/internal/releases/0.0.20"
-  runtime_tests_version: "@unison/runtime-tests/releases/0.0.1"
+  jit_version: "@unison/internal/releases/0.0.21"
```
just curious why did the jit version change?

dolio replied:
I implemented the new serialization format for the jit. This means it now has access to the cacheability information as well.

I was going to implement the top level value behavior in the jit, too, but after some struggling, I realized that my `stronglyConnectedComponents` implementation does not actually output a topological sorting of the SCCs, which I need to output the racket definitions in an acceptable order. So I'll probably just merge this as is.

aryairani replied:
I guess I wouldn't have expected it to change in an interpreter-related PR, but is it just a compatibility thing? That the interpreter was updated to match the new serialization format?

dolio replied:
The interpreter, jit and ucm use a common format for code interchange, so when that changes (to include the cacheability information), the jit needs to be updated to understand it. Otherwise ucm could send it stuff that would just make it bomb. For instance, the format is used for `run.native` and `compile.native` to tell the executable what code to run/compile.

Pretty sure it wouldn't even get past CI if I didn't update it in this case, because the version was incremented.

pchiusano commented:
@SystemFw cool, I'm good with merging this then, whenever @dolio and/or @ChrisPenner are ready.
